src/project-css/app/Libraries/FullTextSearchFormatter.php

```diff
@@ -0,0 +1,139 @@
+// class contains functions that will check and properly Format
+// a string that can be used for full text boolean searches
+// with sql injection in mind.
+
+    private function sanitizeSearchTerm($searchPhrase) {
+       //replace ? with * for wildcard searches
+      $searchPhrase = str_replace('?', '*', $searchPhrase);
+      $searchPhrase = str_replace("+", " +", $searchPhrase);
+      $searchPhrase = str_replace("-", " -", $searchPhrase);
+      $searchPhrase = str_replace("@", "", $searchPhrase);
+      $searchPhrase = str_replace("  ", " ", $searchPhrase);
+      return $searchPhrase;
+    }
+
+    // used by array_filter below filters out items less than 2 characters
+    public function testLength($string) {
+      return (strlen($string) > 1);
+    }
+
+    /**
+     * Tries to clean search string of extra spaces also uses strip_tags
+     * to safeguard AGAINST sql injection. Also replaces ? that is
+     * sometimes used as a wildcard and replaces it with a *
+     * @param       string  $search    Input string
+     * @return      string
+     */
+    public function cleanSearchString($pattern, $search) {
+      // remove excess white space
+      $searchTerms = strip_tags(trim($search));
+      // remove all chars except for letters, numbers and space
+      $value = preg_replace($pattern, '', $searchTerms);
+      $searchTerms = $this->sanitizeSearchTerm($value);
+      $searchArray = (explode(' ', $searchTerms));
+
+      // check remove items with 0 length
+      $searchArray = array_values(array_filter(array_map('trim', $searchArray), array($this, 'testLength')));
+
+      // leave only unique items in array
+      $searchArray = array_unique($searchArray);
+
+      $str = implode(' ', (new customStringHelper)->removeCommonWords($searchArray));
+      return strtolower($str);
+    }
+
+    public function hasRelevancyModifier($word) {
+      if (!empty($word)) {
+        $count = substr_count('<>', $word[0]);
+        if ($count == 1) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    public function hasExactTerm($string) {
+      $count = substr_count($string, '"');
+      if ($count == 2) {
+        return true;
+      }
+      return false;
+    }
+
+    public function hasMatchEither($string) {
+      $startCount = substr_count($string, '(');
+      $endCount = substr_count($string, ')');
+      if (($startCount == 1) && ($endCount == 1)) {
+        return (strpos($string, '(') < strpos($string, ')'));
+      }
+      return false;
+    }
+
+    public function hasWildCard($string) {
+      if (mb_substr($string, -1) == '*') {
+        return true;
+      }
+      return false;
+    }
+
+    public function getMatchEither($string) {
+      // returns string between ()
+      $startPos = strpos($string, '(') + 1;
+      $endPos = strpos($string, ')') - $startPos;
+      return substr($string, $startPos, $endPos);
+    }
+
+    public function cleanMatchEither($string) {
```

Comment:
This is an example, There are 5 logical complexities in this method alone.  I can understand 3, but I think the more logical complexities in the code cause more bugs and showcase a lack of Test Driven Development.  In TDD you wouldn't have made it that far before refactoring these logical complexities out prior to creation.

Reply:
Simplified code for cleanMatchEither. Removed some code for cleaning an array element into a separate function. removed duplicate check on variable.